### PR TITLE
DR-2702 Fix queries for relationship walking to be less CPU heavy

### PR DESCRIPTION
**Pull request by nmalfroy**
Before jumping into the query bits note: as part of the PR I added a logging statement to the big query executions to help future us to debug queries (no data gets logged...just the queries and parameters)

The main thing with this PR is to change how the queries that walk the relationship tree when loading rowids do their joins.  The rough change is that queries that used to look like this:
```
SELECT DISTINCT '<table_id>' AS datarepo_table_id, T.datarepo_row_id
    FROM 
        TO_TABLE T, 
        FROM_TABLE F
WHERE 
    EXISTS (SELECT 1
    FROM UNNEST(F.from_id_field) AS flat_from
    WHERE flat_from = T.to_id_field)
```

now look like
```
SELECT DISTINCT '<table_id>' AS datarepo_table_id, T.datarepo_row_id
    FROM 
        TO_TABLE T, 
        (SELECT F0.*, flat_from
           FROM
             FROM_TABLE F0,
             UNNEST(F0.from_id_field) AS flat_from) F
   WHERE  F.flat_from = T.to_id_field
```

This is an example of the case where the **from** column in the relationship is an array.  The case where the **to** column is an array would look like:
```
SELECT DISTINCT '<table_id>' AS datarepo_table_id, T.datarepo_row_id
    FROM 
        (SELECT F0.*, flat_to
           FROM
             TO_TABLE T0,
             UNNEST(F0.to_id_field) AS flat_to) T, 
        FROM_TABLE F
   WHERE  F.from_id_field = T.flat_from
```

The case where **none** are array would look like:
```
SELECT DISTINCT '<table_id>' AS datarepo_table_id, T.datarepo_row_id
    FROM 
        TO_TABLE T, 
        FROM_TABLE F
   WHERE  F.from_id_field = T.to_id_field
```

And finally, the case where **both** are arrays would look like:
```
SELECT DISTINCT '<table_id>' AS datarepo_table_id, T.datarepo_row_id
    FROM 
        (SELECT F0.*, flat_to
           FROM
             TO_TABLE T0,
             UNNEST(F0.to_id_field) AS flat_to) T, 
        (SELECT F0.*, flat_from
           FROM
             FROM_TABLE F0,
             UNNEST(F0.from_id_field) AS flat_from) F
   WHERE  F.flat_to = T.flat_from
```

### CODE REVIEW

File: src/main/java/bio/terra/service/tabulardata/google/bigquery/BigQuerySnapshotPdao.java [hunk 1]
```diff
@@ -1092,16 +1072,58 @@ private void storeRowIdsForRelatedTable(
...
+              .add("toOrFrom", "TO")
+              .add("field", relationship.getToColumnName());
+      fromCol = relationship.getFromColumnName();
+      toCol = "FLAT_TO";
+    } else {
+      fromTableTableSelect = new ST(tableSelectNonArray);
+      toTableTableSelect = new ST(tableSelectNonArray);
+      fromCol = relationship.getFromColumnName();
+      toCol = relationship.getToColumnName();
+    }
```
Comment (okotsopoulos) [inline src/main/java/bio/terra/service/tabulardata/google/bigquery/BigQuerySnapshotPdao.java]: Is there a reason that you've handled these assignments in this combinatorial way?

I think that handling all the possible from-to combinations in assignment overly complicates things and repeats code.

What about ternary assignments?
```
ST fromTableTableSelect =
  (relationship.getFromColumnIsArray()) ?
  new ST(tableSelectArray).add("toOrFrom", "FROM").add("field", relationship.getFromColumnName()) : 
  new ST(tableSelectNonArray);
String fromCol =
  (relationship.getFromColumnIsArray()) ?
  "FLAT_FROM" : 
  relationship.getFromColumnName();
```

Or handling all the to-or-from assignments in their own block?
```
if (relationship.getFromColumnIsArray()) {
  fromTableTableSelect = …
  fromCol = …
} else {
  fromTableTableSelect = …
  fromCol = …
}
```

**Review by okotsopoulos — APPROVED** [review]
Thank you for the comprehensive examples in the description, and for adding in query / param logging!

I had one suggestion but it is not a blocker to merge.

**Comment by nmalfroy** [comment]
Ah touché.  That's what I get for doing a replacement (the combinatorial approach was more appropriate before) you're right...I like your version a lot more.  I'll update it

**Review by samanehsan — APPROVED** [review]
Lgtm! 💯

**Pull request merged**